src/librustc_typeck/check/expr.rs

```diff
@@ -158,12 +158,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let ty = self.check_expr_kind(expr, expected, needs);
 
         // Warn for non-block expressions with diverging children.
-        match expr.node {
-            ExprKind::Block(..) |
-            ExprKind::Loop(..) | ExprKind::While(..) |
-            ExprKind::Match(..) => {}
-
-            _ => self.warn_if_unreachable(expr.hir_id, expr.span, "expression")
+        if let ExprKind::Block(..) | ExprKind::Loop(..) | ExprKind::Match(..) = expr.node {
```

Comment:
I liked the previous version over having an empty `if` body

Reply:
I really wanted `if !let` for this... and we are inches away from making that a reality implementation-wise. Too bad... I'll revert to a `match`.. :)

Reply:
Done.